### PR TITLE
Add more logging when --verbose is set

### LINKED ISSUE #362: --verbose does not seem to work
Running `notion --verbose install node 8` does not include any information.

I would assume that it would print out things like:

- The version that got resolved (ie 8 -> 8.11.2) 
- The url (and status code) of where notion is attempting to download from
- Where is notion placing node

### LINKED ISSUE #424: Improve output for installing a package
The current process of installing a tool in the default toolchain is noisy—outputting all of npm or yarn's output—and also could be more informative. Suggestions we landed on in our last meeting:

- default to *not* showing npm or yarn's output; instead, only show it if `--verbose` passed, and display a spinner while the process is in progress
- list the installed binaries, e.g. when `volta install ember-cli` finishes say something to the effect of `success: you can now run ember` (preferably with some nice coloring and bolding of the command name or some such)

### PR DESCRIPTION
**Pull request by charlespierce**
Closes #424 
Closes #362 

Info
-----
* Currently the `--verbose` flag when running a Volta command is vastly under-utilized, only affecting the output of errors
* To line up with the expectations of a `--verbose` flag on a CLI, we should output additional information about what Volta is doing at each step of the process.
* Additionally, we are currently showing the output of `npm install` when installing a package binary, but that is noisy and more of an implementation detail. We should only show that output when the user passes `--verbose`

Changes
-----
* Added the `log` crate, which provides macros for logging at different severity levels (`error!`, `warn!`, `info!`, `debug!`, and `trace!`)
* Implemented a custom logger that handles the Volta style for prefixing errors and warnings, as well as a `[verbose]` prefix for all debug output.
* Added initialization of the logger to both `volta` and `shim` binaries, based on the `--verbose` flag and the environment variables.
* Updated the existing calls to `println!` and `eprintln!` to use the logging macros instead.
* Added additional calls to `debug!` with diagnostic information throughout the execution of Volta commands.
* Updated the error reporting to use a function `report` and the log macros, since they already handle which 
* Removed the old `VOLTA_DEV` environment variable in favor of `VOLTA_LOGLEVEL`, which allows you to specifically set the log level you would like to see.

Notes
-----
* Though it's beyond the scope of this PR, the `log` crate abstraction also provides us with the opportunity to have a true log file of everything that happens through Volta.
* Did some small refactors in the `style` crate.

<details><summary>Tasks</summary>

- [X] Add `log` crate as a dependency
- [X] Create custom Logger implementation that handles our Error and Warning prefixes
- [X] Initialize the custom logger in both `volta` and `shim` executables
- [X] Update uses of `println!` & `eprintln!` macros to instead use logging macros
- [X] Add support for `--quiet` as an alternative option, mutually exclusive with `--verbose`
- [x] Support environment variable `VOLTA_LOGLEVEL` to tune the specific supported log level (also useful for debug info on shims)
- [X] Update `volta install` logic to only show the `npm` output if `--verbose` is set.
- [x] Add extra `debug!` statements throughout operations, to allow more information to be shown when a user runs with `--verbose`
</details>

Screenshots
-----
<details><summary>volta install node --verbose</summary>

volta install node --verbose
</details>

<details><summary>volta install yarn --verbose</summary>

volta install yarn --verbose
</details>

<details><summary>volta install ember-cli</summary>

volta install ember-cli
</details>

<details><summary>volta install ember-cli --verbose</summary>

volta install ember-cli --verbose
</details>

<details><summary>volta uninstall ember-cli</summary>

volta uninstall ember-cli
</details>

<details><summary>volta install typescript</summary>

volta install typescript
</details>

<details><summary>volta install typescript --verbose</summary>

volta install typescript --verbose
</details>

<details><summary>volta uninstall typescript</summary>

volta uninstall typescript
</details>

### CODE REVIEW

**Comment by rwjblue** [comment]
In the screenshots when the verbose option is in use, do the stdout / stderr get streamed or do we buffer the content until the command is completed? 

It would be great to stream the content, but I’m not sure how feasible that is.

**Comment by charlespierce** [comment]
@rwjblue Good question! They’re buffered and displayed at the end of the operation. While it’s running we display a spinner describing what is going on (when I get closer to finished with this I’ll try to grab some in-progress screenshots or a video of what it looks like).

I agree it’d be great to stream them, and I actually spent a good chunk of this afternoon trying to figure out a way. Unfortunately it doesn’t seem like the primitives provided by the Rust `std::process::Command` are quite up to the challenge. It seems like even when you connect to the stream there’s some amount of buffering that goes on, and there’s also no way I could find to combine the two streams into a single output in the same way that the terminal does.

The former issue might just be my lack of experience with Rust’s `Read` trait, so if anyone (@dherman, @chriskrycho) has experience with reading a stream in quasi-real time, I’d love to learn more. The latter question is harder, but we might be able to multithread the stream reading and get it done (another topic I know very little about).

**Comment by chriskrycho** [comment]
To do a *proper* streaming API, you'd need to pull in something built on Futures, and I would not recommend that at this point – certainly not till the async/await language primitives stabilize, presumably in 1.37, but even then… probably overkill for this. These two SO Q/As provide some pointers in the right direction, for using a `BufReader` (which was my first thought, but it's not *trivial* because it still depends on how the underlying stream works):

- <https://stackoverflow.com/questions/34611742/how-do-i-read-the-output-of-a-child-process-without-blocking-in-rust>
- <https://stackoverflow.com/questions/31576555/unable-to-pipe-to-or-from-spawned-child-process-more-than-once/31577297#31577297>

**Comment by rwjblue** [comment]
FWIW, the reason I asked RE: streaming is that its probably fairly likely that `--verbose` would be used in scenarios where something is going wrong, for example the process is "hanging". In that circumstance streaming would at least allow the end user to see what `volta` thinks it is doing (e.g. its fetching, its installing npm deps, etc).

Again, I don't think this is a massive issue or anything, but my debugging experience (e.g. with things like `DEBUG=* npm i`) would suggest that at least **sometime** we will want this.

**Comment by chriskrycho** [comment]
Yeah, that's helpful. And it *should* be possible to do that with `BufReader`, though we'll need to experiment and confirm that – how these streams work varies per-OS, and it doesn't *seem* that Rust's primitives here abstract over that (as I'd expect, given Rust's aim).

**Comment by charlespierce** [comment]
@rwjblue Re-reading this thread as I come back to this, I think there might be a slight disconnect here. In the screenshots the `[INSTALL STDOUT]` and `[INSTALL STDERR]` sections are _only_ the output from running the child process `npm install` to get a package dependencies. And it's only the output from that child process that is buffered.

The descriptions of what Volta itself is doing are what I listed as "Add extra `debug!` statements" in the first comment, which I am currently working on. All of those statements will go out in real-time. So you'll see `Fetching package metadata from URL` and similar on the console as it happens, the only one that will be buffered is we'll show `Installing package dependencies` and then it will wait until `npm install` completely finishes to show the output from npm, before continuing with the rest of the process.

**Comment by charlespierce** [comment]
No longer WIP! Added verbose output throughout the fetch and install process.

File: crates/volta-core/src/distro/package.rs [hunk 1]
```diff
@@ -354,7 +379,23 @@ impl PackageVersion {
...
+                debug!(
+                    "Found 'engines.node' specification for {}: {}",
+                    tool_version(&self.name, &self.version),
+                    &engines
+                );
+                engines
+            }
+            None => {
+                debug!(
+                    "No 'engines.node' found for {}, using latest",
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/package.rs]: It's currently the pure latest, but I think there is room for the logic to be adapted. In particular, I think we should try to use the user's default version if it matches the engines (or if no engines are available) first, and only if that doesn't exist fall back to an overall default (which I think makes sense to use LTS).

File: crates/volta-core/src/distro/node.rs [hunk 2]
```diff
@@ -124,10 +125,16 @@ impl NodeDistro {
         let distro_file = path::node_inventory_dir()?.join(&distro_file_name);
 
         if let Some(archive) = load_cached_distro(&distro_file) {
+            debug!(
+                "Loading node from cached archive at {}",
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/node.rs]: Good point! I'll update!

File: crates/volta-core/src/distro/node.rs [hunk 3]
```diff
@@ -124,10 +125,16 @@ impl NodeDistro {
...
+                distro_file.display()
+            );
             return Ok(NodeDistro { archive, version });
         }
 
         ensure_containing_dir_exists(&distro_file)?;
+        debug!("Downloading node from {}", url);
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/node.rs]: Same!

File: crates/volta-core/src/distro/node.rs [hunk 4]
```diff
@@ -153,6 +160,7 @@ impl Distro for NodeDistro {
                 distro: Some(ref hook),
                 ..
             }) => {
+                debug!("Using node.distro hook to determine download URL");
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/node.rs]: At this point, we don't actually have the information about where the hook came from (though I have a plan for a refactor that will help support that as well as the `volta list` command). We _do_ however, also have the debug messages about where configs are found (and if multiples are found, that they are being merged) in `hook/mod.rs`

File: crates/volta-core/src/distro/package.rs [hunk 5]
```diff
@@ -521,12 +579,13 @@ impl Installer {
                     "--loglevel=warn",
                     "--no-update-notifier",
                     "--no-audit",
+                    "--color=always",
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/package.rs]: It's related to `--verbose` because we're only showing the output from this `npm` command if the user has specified `--verbose`. And in those cases, we are managing the output, so `npm` detects that it's _not_ a TTY and suppresses all colors. Then when we show it to the user, it's monotone and hard to parse.

File: crates/volta-core/src/distro/yarn.rs [hunk 6]
```diff
@@ -73,10 +74,16 @@ impl YarnDistro {
         let distro_file = path::yarn_inventory_dir()?.join(&distro_file_name);
 
         if let Some(archive) = load_cached_distro(&distro_file) {
+            debug!(
+                "Loading yarn from cached archive at {}",
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/yarn.rs]: We do, and we should show it as above for `node`, good catch!

File: crates/volta-core/src/distro/yarn.rs [hunk 7]
```diff
@@ -73,10 +74,16 @@ impl YarnDistro {
...
+                distro_file.display()
+            );
             return Ok(YarnDistro { archive, version });
         }
 
         ensure_containing_dir_exists(&distro_file)?;
+        debug!("Downloading yarn from {}", url);
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/yarn.rs]: Yep, updated!

File: crates/volta-core/src/distro/yarn.rs [hunk 8]
```diff
@@ -102,6 +109,7 @@ impl Distro for YarnDistro {
                 distro: Some(ref hook),
                 ..
             }) => {
+                debug!("Using yarn.distro hook to determine download URL");
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/yarn.rs]: Same answer here, we can't currently, but I'd like to change that :+1:

File: tests/acceptance/volta_pin.rs [hunk 9]
```diff
@@ -203,9 +205,7 @@ fn pin_node() {
 
     assert_that!(
         s.volta("pin node@6"),
-        execs()
-            .with_status(ExitCode::Success as i32)
-            .with_stdout_contains("[..]pinned node@6.19.62 (with npm@3.10.1066) in package.json")
```
Comment (mikrostew) [inline tests/acceptance/volta_pin.rs]: Seems like by default we should still have some kind of feedback to the user about which version was pinned, without requiring that they set VOLTA_LOGLEVEL to get that info.

File: crates/volta-core/src/error/details.rs [hunk 10]
```diff
@@ -665,12 +665,20 @@ To {action} the packages '{name}' and '{version}', please {action} them in separ
                     formatted=tool_version(name, version)
                 );
 
-                write!(f, "{}\n\n{}", error, fill(&call_to_action, text_width()))
+                let wrapped_cta = match text_width() {
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/error/details.rs]: We can hit the `None` case if the system is unable to detect the size of the terminal (for instance when we aren't in a TTY so there is no terminal). Before, the code was defaulting to an arbitrary value for the terminal width if it wasn't available, but in those cases we really should just omit the wrapping entirely, as it doesn't make sense to force wrapping when piping to a log file, for example.

File: tests/acceptance/volta_pin.rs (hunk 9 above)
Comment (charlespierce) [reply tests/acceptance/volta_pin.rs]: We do by default in a TTY (but the tests are run with subcommands, so it detects that we aren't a TTY). I could see a different approach where we always write that information out, but by default if we're running as a script then the output is limited to only errors.

File: crates/volta-core/src/error/details.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/error/details.rs]: Also, this is actually a much simpler wrapping than is used by `wrap_content` in `log.rs`, so I think it's reasonable to keep them separate.

File: crates/volta-core/src/distro/package.rs [hunk 11]
```diff
@@ -188,6 +189,15 @@ impl Distro for PackageDistro {
 
         let archive = self.load_or_fetch_archive()?;
 
+        let tmp_root = path::tmp_dir()?;
+        let temp = tempdir_in(&tmp_root)
+            .with_context(|_| ErrorDetails::CreateTempDirError { in_dir: tmp_root })?;
+        debug!(
```
Comment (dherman) [inline crates/volta-core/src/distro/package.rs]: Just a suggestion: it feels like details of logging logic distract from the main line of logic in these functions, and it might be nice to extract them out into a helper method so you can inject them as a single line (e.g., `self.log_unpacking()` or `self.log(Verbose::Unpacking)` or something).

File: crates/volta-core/src/distro/package.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dherman) [reply crates/volta-core/src/distro/package.rs]: Then wouldn't it make sense for us to check for TTY and make that decision dynamically? If someone is calling us from a script they won't want to see the colors.

File: crates/volta-core/src/log.rs [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+/// Determines the correct logging level based on the environment
+/// If VOLTA_LOGLEVEL is set to a valid level, we use that
+/// If not, we check the current stdout to determine whether it is a TTY or not
+///     If it is a TTY, we use Info
+///     If it is NOT a TTY, we use Error as we don't want to show warnings when running as a script
+fn level_from_env() -> LevelFilter {
+    match env::var(VOLTA_LOGLEVEL).as_ref() {
+        Ok(l) if l == "off" => LevelFilter::Off,
```
Comment (dherman) [inline crates/volta-core/src/log.rs]: I _think_ you could write:
```rust
Ok("off") => LevelFilter::Off,
// etc
```
but maybe you tried that already.

Comment (charlespierce) [reply crates/volta-core/src/log.rs]: I did try that initially, the issue is that `env::var` returns a `Result<String, VarError>`, so it throws a type error when trying to match against an `&str` literal. And adding something like `"off".into()` then complains that it's not an Enum variant because it has to calculate something to do the match.

File: crates/volta-core/src/distro/package.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/distro/package.rs]: Good point, we can do that easily :+1:

File: crates/volta-core/src/distro/node.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rwjblue) [reply crates/volta-core/src/distro/node.rs]: Perfect! That makes sense. Its tough to grok the flow of these messages just from the GH review (since files are sorted alpha in GH but obviously that isn't how the actual command line flow works).

File: crates/volta-core/src/error/details.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (rwjblue) [reply crates/volta-core/src/error/details.rs]: Awesome, thank you for explaining!

File: crates/volta-core/src/distro/package.rs (hunk 11 above)
Comment (charlespierce) [reply crates/volta-core/src/distro/package.rs]: Hmm, that's a bit of a balancing act. I think there are benefits to having the logging right in with the other code, for ease of searching when something goes wrong. But having them be single lines is nice, so I went with pulling out the ones in this file that fell in the more complicated methods (`fetch` and `install`). In `node.rs` and `yarn.rs`, all of the ones that are in the analogous methods are already single lines, so I left them alone.

File: tests/acceptance/volta_pin.rs (hunk 9 above)
Comment (mikrostew) [reply tests/acceptance/volta_pin.rs]: ah, gotcha - went back and read through that
